plugins/hls-refactor-plugin/src/Development/IDE/Plugin/CodeAction/Extract.hs

```diff
@@ -0,0 +1,268 @@
+--    foo bar = let tmp = bar; bar = 1 in bar
+-- Where bar is used in the rhs of, but is also bound in the same expression. Thankfully this would just cause an
+-- unbound variable error post-extraction.
+--
+-- In order to have a more robust strategy without excessive maintenance burden, we would require output from
+-- GHC's renamer phase, preferably in-tree.
+--
+-- Some known weaknesses:
+-- TODO Some subtle name shadowing issues in the extracted expression. Hard to fix without more GHC help.
+--      In this case, bar is actually a free variable, but this function will find it not to be free.
+-- TODO Record `{..}` pattern matching syntax; we can find the fields in the matching record to fix this. See
+--      hls-explicit-refcord-fields-plugin
+findFreeVarsPostExtract :: (Monad m) => SrcSpan -> LHsExpr GhcPs -> ParsedSource -> TransformT m [RdrName]
+findFreeVarsPostExtract extractSpan extractExpr parsedSource = do
+      -- Find the variables that are bound in this declaration. Not finding the surrounding decl is a bug.
+      -- TODO return error when we cannot find the containing decl
+      (fromMaybe [] -> boundVarsInWholeDecl) <- querySmallestDeclWithM (pure . (extractSpan `isSubspanOf`)) (pure . boundVarsQ) parsedSource
+      let boundVarsInExtractedExpr = boundVarsQ extractExpr
+      let usedVarsInExtractedExpr = usedVarsQ extractExpr
+      let varsThatScopeOverExtractedExpr =
+            Set.fromList boundVarsInWholeDecl `Set.difference` Set.fromList boundVarsInExtractedExpr
+      -- We use intersection, which implicitly assumes that all non-bound variables in the surrounding scope are
+      -- imports. This is where problems with record field wildcards arise.
+      let postExtractFreeVars = varsThatScopeOverExtractedExpr `Set.intersection` Set.fromList usedVarsInExtractedExpr
+      let argsList = Set.toList postExtractFreeVars
+      pure argsList
+
+-- | A type-safety-helpful newtype that represents that a Range has had its whitespace padding removed from both sides:
+--
+-- Original range:
+--            1 + 1
+--      ^               ^
+--
+-- UnpaddedRange:
+--            1 + 1
+--            ^   ^
+-- See `removeWhitespacePadding` for more info
+newtype UnpaddedRange = UnpaddedRange {unUnpaddedRange :: Range}
+
+-- | This function removes whitespace padding from the Range provided by the user.
+--
+-- When the user selects their range that they would like to extract, we would like to ignore whitespace in their
+-- selection, so that we can find expressions that are contained fully by said range. Consider the following selection:
+--
+--   foo = 1    +    2
+--           ^       ^
+-- In this case, while the range fully wraps `2`, it is not a valid extract function query, since it also wraps `+ 2`,
+-- which is not an expression.
+--
+-- Therefore, in order to catch this case, we narrow the selection to:
+--
+--   foo = 1    +    2
```

Comment:
I'm confused. Now it looks like we've narrowed it to `+ 2`, but you just said that isn't an expression, so how does that help?